### PR TITLE
add plugin for generic keycloak component

### PR DESCRIPTION
**Pull request by fivetide**
##### SUMMARY
This MR adds a new plugin for generic management of keycloak components. It is derived from the keycloak_realm_keys plugin, but omits the key-specific handling of secrets and certs. Instead it is possible to manage any component.
Possible examples are ThemeSelector, LocaleSelector, UserStorageProvider. 
see the official docs for more information:
https://www.keycloak.org/docs/latest/server_development/index.html#_providers

##### ISSUE TYPE
- New Module/Plugin Pull Request

##### COMPONENT NAME
keycloak_component

### CODE REVIEW

File: changelogs/fragments/8826-keycloak-component.yml
Comment (felixfontein) [inline changelogs/fragments/8826-keycloak-component.yml:1]: Please remove the changelog fragment. New modules shouldn't have one (see https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins).

File: plugins/modules/keycloak_component.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    - The names of module options are snake_cased versions of the camelCase ones found in the
+      Keycloak API and its documentation at U(https://www.keycloak.org/docs-api/latest/rest-api/index.html).
+      Aliases are provided so camelCased versions can be used as well.
+
+attributes:
+    check_mode:
+        support: full
+    diff_mode:
+        support: partial
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:35]: You should have `details` here to explain why this is only partial and not full.
```suggestion
        support: partial
        details:
            - ...
```

File: plugins/modules/keycloak_component.py [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+
+options:
+    state:
+        description:
+            - State of the keycloak component.
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:40]: The official spelling is Keycloak with upper-case K.
```suggestion
            - State of the Keycloak component.
```

File: plugins/modules/keycloak_component.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        type: str
+        required: true
+    provider_id:
+        description:
+            - The name of the "provider ID" for the key.
+        type: str
+        required: true
+    provider_type:
+        description:
+            - The name of the "provider type" for the key. I.e. org.keycloak.storage.UserStorageProvider, org.keycloak.userprofile.UserProfileProvider, ...
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:63]:
```suggestion
            - The name of the "provider type" for the key. That is, V(org.keycloak.storage.UserStorageProvider), V(org.keycloak.userprofile.UserProfileProvider), ...
```
See https://docs.ansible.com/ansible/devel/dev_guide/style_guide/index.html#avoid-using-latin-phrases and https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#semantic-markup-within-module-documentation

File: plugins/modules/keycloak_component.py [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+            - see https://www.keycloak.org/docs/latest/server_development/index.html#_providers
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:64]:
```suggestion
            - See U(https://www.keycloak.org/docs/latest/server_development/index.html#_providers).
```

File: plugins/modules/keycloak_component.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+      cachePolicy: "NO_CACHE"
+      enabled: true
+'''
+
+RETURN = '''
+msg:
+    description: Message as to what action was taken.
+    returned: always
+    type: str
+
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:103]: `msg` is one of the standard return values and doesn't need extra mentions.
```suggestion
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Please reference to https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins for how to fix the BOTMETA issue (you need to add the module and yourself to `.github/BOTMETA.yml`).

File: plugins/modules/keycloak_component.py (hunk 1 above)
Comment (fivetide) [reply plugins/modules/keycloak_component.py]: yeah that was copypasta and i didnt give it a second thought. should be 'full'

File: plugins/modules/keycloak_component.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+        required: true
+    provider_id:
+        description:
+            - The name of the "provider ID" for the key.
+        type: str
+        required: true
+    provider_type:
+        description:
+            - The name of the "provider type" for the key. That is, V(org.keycloak.storage.UserStorageProvider),
+            V(org.keycloak.userprofile.UserProfileProvider), ...
```
Comment (felixfontein) [inline plugins/modules/keycloak_component.py:64]: This likely causes the docs parsing error:
```suggestion
              V(org.keycloak.userprofile.UserProfileProvider), ...
```

Comment (fivetide) [reply plugins/modules/keycloak_component.py]: thanks!

File: changelogs/fragments/8826-keycloak-component.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - keycloak_component - add plugin for managing any keycloak component (https://github.com/ansible-collections/community.general/pull/8826)
```
Comment (russoz) [inline changelogs/fragments/8826-keycloak-component.yml:2]: Please use punctuation here, this is later exported to public documentation.
```suggestion
  - keycloak_component - add plugin for managing any keycloak component (https://github.com/ansible-collections/community.general/pull/8826).
```

File: plugins/modules/keycloak_component.py [hunk 8]
```diff
@@ -0,0 +1,330 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2017, Eike Frost <ei@kefro.st>
+# Copyright (c) 2021, Christophe Gilles <christophe.gilles54@gmail.com>
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:5]: You should update this ;-)

File: plugins/modules/keycloak_component.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+short_description: Allows administration of Keycloak components via Keycloak API
+
+version_added: 9.5.0
+
+description:
+    - This module allows the administration of Keycloak components via the Keycloak REST API. It
+      requires access to the REST API via OpenID Connect; the user connecting and the realm being
+      used must have the requisite access rights. In a default Keycloak installation, admin-cli
+      and an admin user would work, as would a separate realm definition with the scope tailored
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:24]: I think using the markup here might be helpful for the readers:
```suggestion
      used must have the requisite access rights. In a default Keycloak installation, C(admin-cli)
      and an C(admin) user would work, as would a separate realm definition with the scope tailored
```

File: plugins/modules/keycloak_component.py [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+    provider_type:
+        description:
+            - The name of the "provider type" for the key. That is, V(org.keycloak.storage.UserStorageProvider),
+              V(org.keycloak.userprofile.UserProfileProvider), ...
+            - See U(https://www.keycloak.org/docs/latest/server_development/index.html#_providers).
+        type: str
+        required: true
+    config:
+        description:
+            - Dict specifying the key and its properties. Contents vary depending on the provider type.
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:70]: A bit odd to define the dict as being a dict. ;-) If I may suggest:
```suggestion
            - Configuration properties for the provider.
            - Contents vary depending on the provider type.
```
Or something along those lines.

File: plugins/modules/keycloak_component.py [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+from ansible.module_utils.six.moves.urllib.parse import urlencode
+from copy import deepcopy
+
+
+def main():
+    """
+    Module execution
+
+    :return:
+    """
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:146]: IMO the code is cleaner without this.

File: plugins/modules/keycloak_component.py [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+
+    argument_spec.update(meta_args)
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True,
+                           required_one_of=([['token', 'auth_realm', 'auth_username', 'auth_password']]),
+                           required_together=([['auth_realm', 'auth_username', 'auth_password']]))
+
+    # Initialize the result object. Only "changed" seems to have special
+    # meaning for Ansible.
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:168]: A bit redundant to describe how module results look like. And, to be fair, `diff` also has special meaning for Ansible.
```suggestion
    # Initialize the result object. Only "changed" seems to have special
    # meaning for Ansible.
```

**Review by russoz — COMMENTED** [review]
hi @fivetide , thanks for your contribution!

Overall the code loogs good, I left just a couple of comments below, and as pointed before you need to add an entry for this module in the `.github/BOTMETA.yml` file.

File: changelogs/fragments/8826-keycloak-component.yml (hunk 7 above)
Comment (felixfontein) [reply changelogs/fragments/8826-keycloak-component.yml:2]: This file needs to be removed anyway.

File: plugins/modules/keycloak_component.py [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: keycloak_component
+
+short_description: Allows administration of Keycloak components via Keycloak API
+
+version_added: 9.5.0
```
Comment (russoz) [inline plugins/modules/keycloak_component.py:17]: 9.5.0 has been released, this must change to:
```suggestion
version_added: 10.0.0
```

**Review by russoz — COMMENTED** [review]
hi @fivetide thanks for your contribution.

Other than Felix's comment, please also update the version number.

Other than that, as far as I can tell, it LGTM

**Comment by felixfontein** [comment]
@fivetide please remove the changelog fragment (see my comment https://github.com/ansible-collections/community.general/pull/8826/files#r1747608052 or the CONTRIBUTING.md file).

**Review by russoz — APPROVED** [review]
LGTM

**Comment by fivetide** [comment]
thanks @felixfontein @russoz for helping with this <3

**Comment by felixfontein** [comment]
FYI @fivetide you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746

**Pull request merged**